### DISCUSSION TITLE
Custom-modules

### DISCUSSION THREAD

**Discussion by DeFlanko** [codeproject/CodeProject.AI-Server#155] [Q&A] [answered]
Is there a centralized custom-modules location?

**Comment by MikeLud** [comment, marked as answer]
> Is there a centralized custom-modules location?

Module links can be found here https://codeproject.github.io/

My custom models can be found here https://github.com/MikeLud/CodeProject.AI-Custom-IPcam-Models

Also you can find custom models here https://huggingface.co/models?pipeline_tag=object-detection&sort=trending&search=yolo

**Reply by 8666** [reply]
I see you've updated the Coral modules.. how can be added to CPAI? Looks like the ObjectDetectionCoral has hardcoded list of modules and does not detect yours even if I add it in the assets subfolder with the rest of .tflite files

**Reply by finnaGIT** [reply]
if you've got the .tflite files in the right location you should be able to make an api call to the custom endpoint: ip:32168/v1/vision/custom/modelname

**Reply by 8666** [reply]
curl -X POST ai:32168/v1/vision/custom/ipcam-general

{"success":false,"error":"Error occurred on the server","moduleId":"ObjectDetectionCoral","moduleName":"Object Detection (Coral)","code":500,"command":"custom","requestId":"38eed4d4-a1e0-4c64-a06d-69a2af0f0659","inferenceDevice":"Multi-TPU","analysisRoundTripMs":3,"processedBy":"localhost","timestampUTC":"Fri, 06 Dec 2024 11:25:08 GMT"}


I've put these under 
/modules/ObjectDetectionCoral/custom-models/ipcam-general.tflite
and also in
/modules/ObjectDetectionCoral/assets/ipcam-general.tflite

I think there is a fixed list in the python code and does not except other models @MikeLud

**Reply by finnaGIT** [reply]
have you gone into the Coral module's options.py and updated the `self.MAX_IDLE_SECS_BEFORE_RECYCLE       = 600000` parameter? by default it times out the multi-tpu after 60s which caused me some issues until i changed it to 600000 lol

**Reply by ChrisMaunder** [reply]
@8666 calling an API without providing the correct payload (in this case an image) will result in an error. This is by design.

**Reply by 8666** [reply]
OK so agentDVR custom module detection is wrong (but I also get the same list on the explorer.html test page). I need to paste "/v1/vision/custom/ipcam-general" in the agentDVR text box and it works. Now I can finally use both Coral Edge TPU (USB) that takes 16-20ms per detection and Yolo8 CPU that takes 115-130ms in parallel and expect the same results.

Can we expect iGPU support on Yolo?

**Reply by mirfan27** [reply]
When working on platforms like [Homeworkify ](https://homeworkifyy.co.uk/)or any custom web application, organizing your codebase is crucial for maintainability and scalability. A centralized custom-modules location is a common practice to keep your project structured and efficient. Here's how it applies:

1. Why Centralize Custom Modules?
Centralizing custom modules ensures that all reusable components, features, or functionalities (like those for Homeworkify) are stored in one place, making them easier to manage, update, and debug.
For example, if you’re building a feature for Homeworkify that handles user authentication or assignment tracking, keeping it in a centralized location ensures consistency across your application.

2. Common Locations for Custom Modules
Drupal: Custom modules are typically stored in /sites/all/modules/custom/ or /sites/[site-name]/modules/custom/.
WordPress: Custom functionality (plugins or modules) is stored in /wp-content/plugins/. For Homeworkify-specific features, you can create a dedicated plugin folder here.
Custom Applications: In frameworks like Laravel or Node.js, you can create a /modules/ or /app/CustomModules/ directory to house all custom modules.

3. Example for a Platform Like Homeworkify
If you’re developing a platform similar to [Homeworkify](https://homeworkifyy.co.uk/), you might structure your custom modules like this:
Copy
/src/modules/  
├── homework-tracking/  
├── user-authentication/  
├── payment-integration/  
└── analytics/ 
This structure ensures that each module is self-contained and can be reused or updated independently.

4. Best Practices
Consistency: Use a clear naming convention for your modules, especially if they relate to Homeworkify features.
Documentation: Document each module’s purpose and functionality for easier onboarding and maintenance.
Version Control: Use Git or another version control system to track changes to your custom modules.

**Comment by DeFlanko** [comment]
Follow up question for you @MikeLud , can any custom models be applied to the License Plate Reader (3.3.0) component? This way it can use those custom modules and still display the licence plate value if detected?

**Reply by ChrisMaunder** [reply]
The ALPR module (at least the non-RKNN version) first makes a call to the main server with a request for a license plate, and once it gets a region back from that call that contains plates, it then uses an OCR model to read the plates. 

When you suggest using custom models do you mean for the plate detection or for the OCR? There's certainly scope to allow custom models in the plate detector to be used, but that would require coordinating two modules. Not hard, just slightly fiddly.

**Reply by DeFlanko** [reply]
i guess both to try other community models, sounds like the APLR pipeline that has the models and process baked in to the module. if there was way to plug in custom models to the pipe line for the plate detection and OCR i wonder if the community could help enhance this easier.
